nixos/modules/services/home-automation/zwave-js.nix

```diff
@@ -0,0 +1,150 @@
+{config, pkgs, lib, ...}:
+
+with lib;
+
+let
+  cfg = config.services.zwave-js;
+  cacheDir = "/var/cache/zwave-js";
+  runtimeDir = "/run/zwave-js";
+  defaultConfig = {
+    storage = {
+      inherit cacheDir;
+    };
+  };
+  mergedConfigFile = "${runtimeDir}/config.json";
+in {
+
+  options.services.zwave-js = {
+    enable = mkEnableOption (mdDoc "Start the zwave-js server on boot");
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.zwave-js-server;
+      defaultText = literalExpression "pkgs.zwave-js-server";
+      description = mdDoc ''
+        Package for zwave-js server
+      '';
+    };
+
+    port = mkOption {
+      type = types.int;
+      default = 3000;
+      description = mdDoc ''
+        Port for the server to listen on.
+      '';
+    };
+
+    serialPort = mkOption {
+      type = types.nullOr types.path;
+      default = null;
+      description = mdDoc ''
+        Serial port device path for Z-Wave controller.
+      '';
+      example = "/dev/ttyUSB0";
+    };
+
+    secretsConfigFile = mkOption {
+      type = types.path;
+      description = mdDoc ''
+        JSON file containing secret keys. This file will be merged with the
+        module-generated config file (taking precedence).
+
+        Z-Wave keys can be generated with:
+
+          {command}`< /dev/urandom tr -dc A-F0-9 | head -c32 ;echo`
+
+        ::: {.warning}
+        A file in the nix store should not be used since it will be readable to
+        all users.
+        :::
+      '';
+      example = "/secrets/zwave-js-keys.json";
+    };
+
+    settings = mkOption {
+      type = types.attrs;
+      default = {};
+      description = mdDoc ''
+        Extra configuration settings to be merged into the generated config
+        file.
+      '';
+      example = literalExpression ''
+        {
+          storage = {
+            throttle = "fast";
+          };
+        }
+      '';
+    };
+
+    mockDriver = mkOption {
+      type = types.bool;
+      default = false;
+      description = mdDoc ''
+        Use a mock USB serial driver (mutually exclusive with serialPort option).
+      '';
+      internal = true;
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.mockDriver && cfg.serialPort == null || !cfg.mockDriver && cfg.serialPort != null;
+        message = "The zwave-js-server module requires one of serialPort or mockDriver but not both";
+      }
+    ];
+    systemd.services.zwave-js = let
+      configFile = pkgs.writeText "zwave-js-config.json" (builtins.toJSON
+        (recursiveUpdate defaultConfig cfg.settings)
+      );
```

Comment:
The values provided through `defaultConfig` are not really discoverable without inspecting this file, which is a design limitation of using this approach.

Ideally the `settings` option would use RFC42 (submodule with freeformType `pkgs.formats.json`) and the default values would be options on that submodule.

Reply:
I've had a go at making this change, let me know what you think.